File: Graph_Functions/Kruskal_s/Kruskal.py
class Edge:
    def __init__(self, u, v, weight):
        self.u = u
        self.v = v
        self.weight = weight

    def __lt__(self, other):
        return self.weight < other.weight

def find_parent(u, parent):
    if parent[u] != u:
        parent[u] = find_parent(parent[u], parent)
    return parent[u]

def union_sets(u, v, parent, rank):
    root_u = find_parent(u, parent)
    root_v = find_parent(v, parent)

    if root_u != root_v:
        if rank[root_u] > rank[root_v]:
            parent[root_v] = root_u
        elif rank[root_u] < rank[root_v]:
            parent[root_u] = root_v
        else:
            parent[root_v] = root_u
            rank[root_u] += 1

def kruskal(n, edges):
    parent = [i for i in range(n)]
    rank = [0] * n
    mst_edges = []

    edges.sort()

    for edge in edges:
        root_u = find_parent(edge.u, parent)
        root_v = find_parent(edge.v, parent)

        if root_u != root_v:
            mst_edges.append(edge)
            union_sets(edge.u, edge.v, parent, rank)

    return mst_edges

def main():
    n = 4
    edges = [
        Edge(0, 1, 10),
        Edge(0, 2, 6),
        Edge(0, 3, 5),
        Edge(1, 3, 15),
        Edge(2, 3, 4),
    ]

    mst_edges = kruskal(n, edges)

    print("Edges in the Minimum Spanning Tree (MST):")
    mst_weight = 0
    for edge in mst_edges:
        print(f"{edge.u} - {edge.v} with weight {edge.weight}")
        mst_weight += edge.weight

    print(f"Total weight of MST: {mst_weight}")

if __name__ == "__main__":
    main()

File: Graph_Functions/Kruskal_s/Kruskal.java
import java.util.*;

class Edge implements Comparable<Edge> {
    int u, v, weight;

    Edge(int u, int v, int weight) {
        this.u = u;
        this.v = v;
        this.weight = weight;
    }

    @Override
    public int compareTo(Edge other) {
        return this.weight - other.weight;
    }
}

public class KruskalAlgorithm {
    public static int findParent(int u, int[] parent) {
        if (parent[u] == u) {
            return u;
        }
        return parent[u] = findParent(parent[u], parent);
    }

    public static void unionSets(int u, int v, int[] parent, int[] rank) {
        int rootU = findParent(u, parent);
        int rootV = findParent(v, parent);

        if (rootU != rootV) {
            if (rank[rootU] > rank[rootV]) {
                parent[rootV] = rootU;
            }
            else if (rank[rootU] < rank[rootV]) {
                parent[rootU] = rootV;
            }
            else {
                parent[rootV] = rootU;
                rank[rootU]++;
            }
        }
    }

    public static List<Edge> kruskal(int n, List<Edge> edges) {
        int[] parent = new int[n];
        int[] rank = new int[n];
        List<Edge> mstEdges = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }

        Collections.sort(edges);

        for (Edge e : edges) {
            int rootU = findParent(e.u, parent);
            int rootV = findParent(e.v, parent);

            if (rootU != rootV) {
                mstEdges.add(e);
                unionSets(e.u, e.v, parent, rank);
            }
        }

        return mstEdges;
    }

    public static void main(String[] args) {
        int n = 4;
        List<Edge> edges = Arrays.asList(
            new Edge(0, 1, 10),
            new Edge(0, 2, 6),
            new Edge(0, 3, 5),
            new Edge(1, 3, 15),
            new Edge(2, 3, 4)
        );

        List<Edge> mstEdges = kruskal(n, edges);

        System.out.println("Edges in the Minimum Spanning Tree (MST):");
        int mstWeight = 0;
        for (Edge e : mstEdges) {
            System.out.println(e.u + " - " + e.v + " with weight " + e.weight);
            mstWeight += e.weight;
        }

        System.out.println("Total weight of MST: " + mstWeight);
    }
}
